### DISCUSSION TITLE
findMany fails due to transaction issue

### DISCUSSION THREAD

**Discussion by DavidLangva** [prisma/prisma#25264] [Q&A]
### Question

When querying large tables with big returns, I sometimes get a findMany failure due to a transaction failure for write conflict or deadlock. 

Seems weird that a write error would come up for a findMany query. Maybe its due to the select statements?

```
Invalid `prisma.lineItem.findMany()` invocation:


Transaction failed due to a write conflict or a deadlock. Please retry your transaction
```

Here is the query that is failing

```ts 
const lineItems = await prisma.lineItem.findMany({
        select: {
          genre: {
            select: {
              name: true,
            },
          },
          image: {
            select: {
              person: {
                select: {
                  name: true,
                },
              },
              places: {
                select: {
                  texts: true,
                },
              },
            },
          },
          name: true,
        },
        take: 5000,
      });
```

**Comment by ludralph** [comment]
Hi @DavidLangva 👋 

According to the [documentation](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#transaction-timing-issues), to resolve this issue on your transaction, you need to set the `isolationLevel` parameter to `Prisma.TransactionIsolationLevel.Serializable`. 

This ensures that your application commits multiple concurrent or parallel transactions as if they were run serially.  In your application code, add a retry around your transaction to handle any P2034 errors. For example:
```js
import { Prisma, PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();
const MAX_RETRIES = 5;

async function fetchLineItemsWithRetry(retries = 0) {
  try {
    const lineItems = await prisma.$transaction(
      [
        prisma.lineItem.findMany({
          select: {
            genre: {
              select: {
                name: true,
              },
            },
            image: {
              select: {
                person: {
                  select: {
                    name: true,
                  },
                },
                places: {
                  select: {
                    texts: true,
                  },
                },
              },
            },
            name: true,
          },
          take: 5000,
        }),
      ],
      {
        isolationLevel: Prisma.TransactionIsolationLevel.Serializable, // added the isolation level
      }
    );
    return lineItems;
  } catch (error) {
    if (error.code === 'P2034' && retries < MAX_RETRIES) {
      console.warn(`Retrying transaction due to conflict or deadlock (attempt ${retries + 1})`);
      return fetchLineItemsWithRetry(retries + 1);
    } else {
      throw error;
    }
  }
}

fetchLineItemsWithRetry()
  .then(lineItems => {
    console.log('Fetched line items:', lineItems);
  })
  .catch(error => {
    console.error('Failed to fetch line items:', error);
  });
```

**Comment by DavidLangva** [comment]
Hi @ludralph and thanks for the reply :-)

interesting! In my case this is happening when running a `prisma.findMany()` and its not inside of a transaction. This is why its especially confusing. Its a read operation run on its own, but errors out with a transaction deadlock. I thought maybe prisma is using a transaction under the hood?

We've also extended the prisma client to use a read replica. Not sure if that could have an implication here.

**Reply by ludralph** [reply]
Hi @DavidLangva 

Thanks for providing more information about what is causing the deadlock. `findMany` doesn't use transactions under the hood. So it is worth checking to see what could be causing this. Have you tried running the same query without the read replica to see if it succeeds? Also, do you have indexes configured for the ready-heavy operation?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
